framework/src/main/java/org/checkerframework/framework/util/QualifierKindHierarchy.java

```diff
@@ -0,0 +1,874 @@
+                    }
+                }
+            }
+            if (qualifierKind.top == null) {
+                throw new UserError("Qualifier isn't in hierarchy: %s", qualifierKind);
+            }
+        }
+        for (QualifierKind qualifierKind : nameToQualifierKind.values()) {
+            for (QualifierKind bot : bottoms) {
+                if (bot.top != qualifierKind.top) {
+                    continue;
+                }
+                if (qualifierKind.bottom == null) {
+                    qualifierKind.bottom = bot;
+                } else if (qualifierKind.top != bot) {
+                    throw new UserError(
+                            "Multiple bottoms found for qualifier %s. Tops: %s and %s.",
+                            qualifierKind, bot, qualifierKind.bottom);
+                }
+                if (qualifierKind.isPoly) {
+                    bot.superTypes.add(qualifierKind);
+                }
+            }
+        }
+    }
+
+    /**
+     * Returns the set of all qualifiers that are a supertype of {@code qualifierKind}.
+     *
+     * @param qualifierKind the qualifier whose super types should be returned
+     * @param directSuperMap directSuperMap
+     * @return the set of all qualifiers that are a supertype of {@code qualifierKind}
+     * @throws UserError if there is a cycle in the hierarchy
+     */
+    private Set<QualifierKind> findAllTheSupers(
+            QualifierKind qualifierKind, Map<QualifierKind, Set<QualifierKind>> directSuperMap) {
+        Queue<QualifierKind> queue = new ArrayDeque<>(directSuperMap.get(qualifierKind));
+        Set<QualifierKind> allSupers = new TreeSet<>(directSuperMap.get(qualifierKind));
+        while (!queue.isEmpty()) {
+            QualifierKind superQual = queue.remove();
+            if (superQual == qualifierKind) {
+                throw new UserError("Cycle in hierarchy: %s", qualifierKind);
+            }
+            queue.addAll(directSuperMap.get(superQual));
+            allSupers.addAll(directSuperMap.get(superQual));
+        }
+        return allSupers;
+    }
+
+    /**
+     * Creates a mapping from {@link QualifierKindPair} to the least upper bound of both
+     * QualifierKinds.
+     *
+     * @return a mapping from {@link QualifierKindPair} to their lub.
+     */
+    protected Map<QualifierKindPair, QualifierKind> createLubsMap() {
+        Map<QualifierKindPair, QualifierKind> lubs = new TreeMap<>();
+        for (QualifierKind qual1 : nameToQualifierKind.values()) {
+            for (QualifierKind qual2 : nameToQualifierKind.values()) {
+                if (qual1.top != qual2.top) {
+                    continue;
+                }
+                QualifierKind lub = findLub(qual1, qual2);
+                QualifierKindPair pair = new QualifierKindPair(qual1, qual2);
+                QualifierKind otherLub = lubs.get(pair);
+                if (otherLub != null) {
```

Comment:
When reading this, I was confused abouh how this can happen.  I think the answer is that QualifierKindPair canonicalizes the order of the elements of the pair.  With the elimination of that class that I suggested, this will not be relevant any more.